docs/specs/pipelines-io.md

```diff
@@ -0,0 +1,196 @@
+    public bool ReadOnly { get; set; }
+    
+    public int ReadableBytes => End - Start;
+    public int WritableBytes => Data.Length - End;
+}
+```
+
+A `BufferSegment` represents a link in a contiguous chain of buffers. New segments are added if the requested memory is less than the writable bytes in the current segment. What's interesting here is that there are 2 views to a `BufferSegment`, the ReadableBytes and the WritableBytes. The writing and reading primitives in pipelines build on this concept.
+
+### Writing
+
+```C#
+private Task Produce(IPipelineWriter writer)
+{
+   while (true)
+   {
+       // Allocate a buffer from the writer
+       WritableBuffer buffer = writer.Alloc();
+       
+       var bytes = Encoding.UTF8.GetBytes(DateTime.Now.ToString());
+       buffer.Write(bytes);
+
+       // Tell the producer we have data
+       await buffer.FlushAsync();
+       
+       // Wait a second before writing again
+       await Task.Delay(1000);
+   }
+}
+```
+
+In the above example, `Produce` calls `Alloc` on the `IPipelineWriter` to get access to the `WritableBuffer`, then it calls `FlushAsync` to signal to the consumer that data is ready. A `WritableBuffer` is a struct that allows writing to an *ever* growing sink of bytes.
+
+```C#
+public struct WritableBuffer
+{
+   public Memory<byte> Memory { get; }
+   public void Ensure(int minimumBytes = 0);
+   public void Advance(int bytesWritten);
+   public void Commit();
+   public Task FlushAsync();
+}
+```
+
+There are extension methods like `Write(ReadOnlySpan<byte>)` which automate writing to the available memory and advancing the buffer.
+
+```C#
+public static void Write(this WritableBuffer buffer, ReadOnlySpan<byte> source)
+{
+    if (buffer.Memory.IsEmpty)
+    {
+        buffer.Ensure();
+    }
+
+    // Fast path, try copying to the available memory directly
+    if (source.Length <= buffer.Memory.Length)
+    {
+        source.CopyTo(buffer.Memory.Span);
+        buffer.Advance(source.Length);
+        return;
+    }
+
+    var remaining = source.Length;
+    var offset = 0;
+
+    while (remaining > 0)
+    {
+        var writable = Math.Min(remaining, buffer.Memory.Length);
+
+        buffer.Ensure(writable);
+
+        if (writable == 0)
+        {
+            continue;
+        }
+
+        source.Slice(offset, writable).CopyTo(buffer.Memory.Span);
+
+        remaining -= writable;
+        offset += writable;
+
+        buffer.Advance(writable);
+    }
+}
+```
+
+### Reading
+
+The consumer calls `IPipelineReader.ReadAsync()` which returns with a `ReadableBuffer` when the producer says that data is ready. 
+
+```C#
+private Task Consume(IPipelineReader reader)
+{
+    while (true)
+    {
+        ReadResult result = await reader.ReadAsync();
+        ReadableBuffer buffer = result.Buffer;
+        if (buffer.IsEmpty && result.IsCompleted)
+        {
+            break;
+        }
+
+        // Process the buffer
+        Consume(ref buffer);
```

Comment:
Why does it have to be passed by ref? Is it because the buffer is a large struct, or it needs to be mutated?